plugins/inputs/opensearch_query/opensearch_query.go

```
@@ -0,0 +1,326 @@
+	now := time.Now().UTC()
+	from := now.Add(time.Duration(-aggregation.QueryPeriod))
+	filterQuery := aggregation.FilterQuery
+	if filterQuery == "" {
+		filterQuery = "*"
+	}
+
+	aq := &Query{
+		Size:         0,
+		Aggregations: aggregation.aggregation,
+		Query:        nil,
+	}
+
+	boolQuery := &BoolQuery{
+		FilterQueryString: filterQuery,
+		TimestampField:    aggregation.DateField,
+		TimeRangeFrom:     from,
+		TimeRangeTo:       now,
+		DateFieldFormat:   aggregation.DateFieldFormat,
+	}
+
+	aq.Query = boolQuery
+	req, err := json.Marshal(aq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %s", err)
+	}
+	o.Log.Debugf("{\"body\": %s}", string(req))
+
+	searchRequest := &opensearchapi.SearchRequest{
+		Body:    strings.NewReader(string(req)),
+		Index:   []string{aggregation.Index},
+		Timeout: time.Duration(o.Timeout),
+	}
+
+	resp, err := searchRequest.Do(ctx, o.osClient)
+	if err != nil {
+		return nil, err
+	}
+	if resp.IsError() {
+		return nil, fmt.Errorf("Opensearch SearchRequest failure: [%d] %s", resp.StatusCode, resp.Status())
+	}
+	defer resp.Body.Close()
+
+	var searchResult AggregationResponse
+
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&searchResult)
+	if err != nil {
+		return nil, err
+	}
+
+	return &searchResult, nil
+}
+
+// getMetricFields function returns a map of fields and field types on Elasticsearch that matches field.MetricFields
+func (o *OpensearchQuery) getMetricFields(ctx context.Context, aggregation osAggregation) (map[string]string, error) {
+	mapMetricFields := make(map[string]string)
+	fieldMappingRequest := opensearchapi.IndicesGetFieldMappingRequest{
+		Index:  []string{aggregation.Index},
+		Fields: aggregation.MetricFields,
+	}
+
+	response, err := fieldMappingRequest.Do(ctx, o.osClient)
+	if err != nil {
+		return nil, err
+	}
+
+	// Bad request; move on
+	if response.StatusCode != 200 {
+		return mapMetricFields, nil
+	}
```

Would it make sense to accept all 2xx error codes?

---

Does this response maybe also have an `IsError()` method as used above?

And shouldn't this method return an error instead of just an empty map?